### PR TITLE
Support alternate Dockerfile name.

### PR DESCRIPTION
**Pull request by KyleJamesWalker**
Add support for compose support Docker's `-f`/`--file` option from within your docker-compose.yml file.

```yaml
app:
  build: .
  dockerfile: Dockerfile-alternate
```

I quickly threw this in, and it seems to work for me, please let me know if anything else is needed to get this into master.

### CODE REVIEW

File: compose/service.py [hunk 1]
```diff
@@ -479,6 +488,7 @@ def build(self, no_cache=False):
             stream=True,
             rm=True,
             nocache=no_cache,
+            fileobj=self._get_fileobj(),
```
Comment (dnephin) [inline compose/service.py]: This isn't quite right, fileobj would be for the whole context.  This needs to use the `dockerfile` kwarg.

https://github.com/docker/docker-py/blob/master/docker/client.py#L301

I don't think that option is in a release yet. I added it after the 1.0.0 release

Comment (KyleJamesWalker) [inline compose/service.py]: Yeah the version I had locally had the following, so I'm assuming it hasn't been released yet
```python
    def build(self, path=None, tag=None, quiet=False, fileobj=None,
              nocache=False, rm=False, stream=False, timeout=None,
              custom_context=False, encoding=None, pull=True,
              forcerm=False):
```

This was the only way I could figure out how to get it to work (with current release), `dockerfile` param would be much better once it's released.

Comment (KyleJamesWalker) [inline compose/service.py]: Do you think I should update this to use dockerfile (even know it's not released yet?)

Comment (dnephin) [inline compose/service.py]: That would be great. I think we can ask for a new release when this is ready.

I believe fileobj doesn't actually work when the Dockerfile makes use of other files in the build context.

Comment (KyleJamesWalker) [inline compose/service.py]: yea I was just testing that, and you are correct.  I'm fighting with getting the client installed to test it out.

Comment (dnephin) [inline compose/service.py]: you can use a git url with pip https://pip.pypa.io/en/latest/reference/pip_install.html#git to get the latest master of docker-py.


As a side note, I wonder if the config option should be `dockerfile`, to keep it consistent with the api  key.

Comment (KyleJamesWalker) [inline compose/service.py]: Within `docker-py` there's a `utils.compare_version`
``` python
def utils.def compare_version(v1, v2):
    """Compare docker versions

    >>> v1 = '1.9'
    >>> v2 = '1.10'
    >>> compare_version(v1, v2)
    1
    >>> compare_version(v2, v1)
    -1
    >>> compare_version(v2, v2)
    0
    """
    s1 = StrictVersion(v1)
    s2 = StrictVersion(v2)
    if s1 == s2:
        return 0
    elif s1 > s2:
        return -1
    else:
        return 1
```

Do you know if there's a helper like this already within compose so I could add something like :
```python
        if 'dockerfile' in options and utils.compare_version('1.17', client.api_version()) is -1:
            raise ConfigError('Docker client API must be greater than 1.17 to support the dockerfile option')
```

and when building:
``` python
    def build(self, no_cache=False):
        log.info('Building %s...' % self.name)

        build_kwars = dict(
            tag=self.full_name,
            stream=True,
            rm=True,
            nocache=no_cache,
        )
        if 'dockerfile' in self.options and \
                utils.compare_version('1.17', self.client.api_version()) is not -1:
            build_kwars['dockerfile'] = self.options.get('dockerfile', None)

        build_output = self.client.build(
            self.options['build'],
            **build_kwars
        )
```

Comment (dnephin) [inline compose/service.py]: There isn't, but I wouldn't worry about it. We're planning on requiring the latest docker to pickup some new features in 1.6 anyway.

**Comment by KyleJamesWalker** [comment]
I haven't tested this with unreleased docker-py, but hopefully it should work.

The `dockerfile` option within the `docker-compose.yml` file will be be checked to maintain compatibly with current `docker-py`

Let me know what you think, thanks for helping with this.

**Comment by KyleJamesWalker** [comment]
Note: Not sure where the api version is coming from I did the following to update `docker-py` to master:
< $ pip install -U git+git://github.com/docker/docker-py.git
```
Collecting git+git://github.com/docker/docker-py.git
  Cloning git://github.com/docker/docker-py.git to /var/folders/1n/0vvgtlpj6sj4tzxn5znyy4jh0000gp/T/pip-_yXZSG-build
    /usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/setuptools-12.0.5-py2.7.egg/setuptools/dist.py:283: UserWarning: The version specified requires normalization, consider using '1.0.1.dev0' instead of '1.0.1-dev'.
Requirement already up-to-date: requests>=2.5.2 in /usr/local/lib/python2.7/site-packages (from docker-py==1.0.1.dev0)
Requirement already up-to-date: six>=1.3.0 in /usr/local/lib/python2.7/site-packages (from docker-py==1.0.1.dev0)
Requirement already up-to-date: websocket-client>=0.11.0 in /usr/local/lib/python2.7/site-packages (from docker-py==1.0.1.dev0)
Installing collected packages: docker-py
  Running setup.py install for docker-py
    /usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/setuptools-12.0.5-py2.7.egg/setuptools/dist.py:283: UserWarning: The version specified requires normalization, consider using '1.0.1.dev0' instead of '1.0.1-dev'.
Successfully installed docker-py-1.0.1.dev0
```

< $ docker-compose up
```
Docker client API must be greater than 1.17 to support the dockerfile option, currently 1.14
```
< $ docker version
```
Client version: 1.5.0
Client API version: 1.17
Go version (client): go1.4.1
Git commit (client): a8a31ef
OS/Arch (client): darwin/amd64
Server version: 1.5.0
Server API version: 1.17
Go version (server): go1.4.1
Git commit (server): a8a31ef
```

**Comment by KyleJamesWalker** [comment]
Just saw your command about the version check (got squished under the change), let me know if you'd like me to remove the checks that I added.

**Comment by dnephin** [comment]
Ya, I would remove it. If someone is an on earlier version, they can just not specify the option and it should continue to work without the dockerfile support.

**Comment by mwean** [comment]
I tried to use this locally, but ran into some issues. I got the error about the API version, so I removed those checks, but now I'm getting `build() got an unexpected keyword argument 'dockerfile'` I believe I have the latest docker version:

```bash
$ docker version
Client version: 1.5.0
Client API version: 1.17
Go version (client): go1.4.1
Git commit (client): a8a31ef
OS/Arch (client): darwin/amd64
Server version: 1.5.0
Server API version: 1.17
Go version (server): go1.4.1
Git commit (server): a8a31ef
```

And I also updated docker-py to the latest commit. Is there anything else I need to do to get this working? Thanks!

**Comment by KyleJamesWalker** [comment]
@mwean You need to install the version from github `pip install -U git+git://github.com/docker/docker-py.git` but without some changed to docker-py the API version detects as 1.15 (but at least it won't crash when trying to pass the argument).  We just have to wait for `docker-py` to be upgraded a bit more.  Also if you install from git you'll need to edit the requirements for `compose`:

From:
```
'requests >= 2.2.1, < 2.5.0',
```

To:
```
'requests >= 2.5.2',
```
otherwise you'll get a version conflict.

**Comment by KyleJamesWalker** [comment]
@dnephin I removed the checks, but would you like me to add the **build_kwargs portion back in.  That way if they are on an outdated version of `docker-py` it won't crash.  Or the requirements just need to be update to make sure a new enough version is required.  Let me know, thanks!

**Comment by bfirsh** [comment]
Maybe this?

```
app:
  build:
    context: .
    file: Dockerfile-alternate
```

**Comment by dnephin** [comment]
I think it's a lot less intuitive if a field can be a string or a mapping. It also makes it a lot harder (maybe impossible) to encode the expected structure of the configuration (in something like http://json-schema.org/ for example).

If we were able to break backwards compatibility, and always do

```
app:
  build:
    context: .
```

and build was never just a string, I think it would be nice.

**Comment by dnephin** [comment]
BTW, there is now a docker-py 1.1.0, so if you update the setup.py:install_requires and the requirements.txt , the tests should pass on docker 1.5+

**Comment by KyleJamesWalker** [comment]
Oh awesome, I'll give it a try today!!!

**Comment by KyleJamesWalker** [comment]
It's not working for me, the container builds, but then exits on up ```
app_1 exited with code 0
```

will look into this more when I have time later, unless someone has an idea, thanks.

**Comment by loic** [comment]
This PR has been working well for me although I had to:
- change the version of `requests` in requirements.txt.
- change the API version given to the `docker-py` client to `1.17`.

**Comment by KyleJamesWalker** [comment]
@Ioic you mean this line?

**Comment by aanand** [comment]
Needs rebasing now that #1100 is merged.

**Comment by KyleJamesWalker** [comment]
Rebased with upstream/master

**Comment by loic** [comment]
@KyleJamesWalker yes that was the line. Maybe consider squashing into a single commit?

**Comment by KyleJamesWalker** [comment]
@loic Didn't know about squashing, hope that did it correctly.  Thanks for the tip!

**Comment by KyleJamesWalker** [comment]
Looks like all the errors are:
> APIError: 404 Client Error: Not Found ("client and server don't have same version (client : 1.17, server: 1.15)")

@aanand do I need to do anything else to get this merged in?

**Comment by aanand** [comment]
If you rebase on master, then remove all the Docker versions apart from 1.6.0-rc2, it should fix the build.

**Comment by KyleJamesWalker** [comment]
@aanand Rebased.

**Comment by aanand** [comment]
Dockerfile still needs updating to remove Docker versions 1.3.3, 1.4.1 and 1.5.0.

**Comment by KyleJamesWalker** [comment]
I removed all `1.3.3, 1.4.1 and 1.5.0` and then just removed `1.3.3 and 1.4.1` and both are failing on the cleanup `rmi` on jenkins.

**Comment by KyleJamesWalker** [comment]
@aanand Do you need me to make any changes for this PR, or does Jenkins need a change for the job?

Also I added you to my fork if you want to make any changes to this PR.

**Comment by aanand** [comment]
The [line that sets the default Docker](https://github.com/docker/compose/blob/db852e14e4d8ebc1f0a477360ceef7694970dc59/Dockerfile#L28-L29) is missing from your Dockerfile and breaking the build. The default was 1.3.3 before - now it needs to be 1.5.0:

```dockerfile
# Set the default Docker to be run
RUN ln -s /usr/local/bin/docker-1.5.0 /usr/local/bin/docker
```

**Comment by KyleJamesWalker** [comment]
Thanks @aanand 

Well at least the tests are running now, I'll take a look when I can to see why a few of the tests are failing.

**Comment by aanand** [comment]
OK, I've updated master so that we're only testing against Docker 1.6, so if you rebase you should be able to discard any changes to the Dockerfile.

Your test failure suggests that docker-py needs updating to 1.1.0 in `requirements.txt`. I have a feeling that's going to break the build when we attempt to compile the binary though - an ongoing issue with the `requests` library, which is being tracked in https://github.com/kennethreitz/requests/issues/2465. Still, give it a try!

**Comment by joost** [comment]
+1 for this PR :)

**Comment by KyleJamesWalker** [comment]
Router died last night so no internet, I'll rebase later today/tonight.

**Comment by atrauzzi** [comment]
+1 would love to be able to start using docker compose! This is the feature preventing it from happening right now. :)

**Comment by dustinlacewell** [comment]
+1 on this PR!

**Comment by KyleJamesWalker** [comment]
Any ideas on how to get the new version of docker-py to play nice with unit tests?

@aanand

**Comment by bfirsh** [comment]
Looks like you need to update requirements.txt to be docker-py 1.1.0.

**Comment by aanand** [comment]
This diff will do it:

```diff
diff --git a/requirements.txt b/requirements.txt
index 4c4113a..f1042fa 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,8 +1,8 @@
 PyYAML==3.10
-docker-py==1.0.0
+docker-py==1.1.0
 dockerpty==0.3.2
 docopt==0.6.1
-requests==2.2.1
+requests==2.6.0
 six==1.7.3
 texttable==0.8.2
 websocket-client==0.11.0
diff --git a/setup.py b/setup.py
index 39ac0f6..f9e5bc8 100644
--- a/setup.py
+++ b/setup.py
@@ -27,10 +27,10 @@ def find_version(*file_paths):
 install_requires = [
     'docopt >= 0.6.1, < 0.7',
     'PyYAML >= 3.10, < 4',
-    'requests >= 2.2.1, < 2.6',
+    'requests >= 2.6.0, < 2.7',
     'texttable >= 0.8.1, < 0.9',
     'websocket-client >= 0.11.0, < 1.0',
-    'docker-py >= 1.0.0, < 1.2',
+    'docker-py >= 1.1.0, < 1.2',
     'dockerpty >= 0.3.2, < 0.4',
     'six >= 1.3.0, < 2',
 ]
```

**Comment by KyleJamesWalker** [comment]
Looking into this error now from Pyinstaller, due to the new requests version
```
ImportError: No module named 'requests.packages.chardet.sys'
```

More info on this here: http://stackoverflow.com/questions/28775276/pyinstaller-error-importerror-no-module-named-requests-packages-chardet-sys

**Comment by aanand** [comment]
Yep, thought that might happen. I'm using a version off GitHub in https://github.com/docker/compose/pull/1269:

    -e git+https://github.com/kennethreitz/requests.git@1b5bfe681b4c0a987e97ae78b2034db7b7ce3d01#egg=requests

**Comment by KyleJamesWalker** [comment]
@aanand Tried adding `-e` to `requirements-dev.txt` with the git sha you mentioned, and it's still failing with the same error.

I also tried adding `-Ur` to the Dockerfile for the dev requirements, but then it failed installing one of the packages with `ImportError: No module named setuptools`.
See: https://jenkins.dockerproject.com/job/Compose-PRs/255/console

I'll try a bit more later today when I have time.

**Comment by nicbarker** [comment]
Can't wait for this PR to get merged, great work @KyleJamesWalker

**Comment by andreausu** [comment]
+1 thank you for working on this, I can't wait for this to be merged!

**Comment by stevenjack** [comment]
+1 be nice to see this go in :)

**Comment by throrin19** [comment]
+1 I can't wait for this

**Comment by jonbrouse** [comment]
+1 this is gonna be awesome

**Comment by jakubriedl** [comment]
+1 cool feature - we are waiting for it

**Comment by aanand** [comment]
requests 2.6.1 is out, so try updating `requirements.txt` with `requests==2.6.1`.

**Comment by KyleJamesWalker** [comment]
Thanks, I'll give it a try when I make it into work today!

**Comment by KyleJamesWalker** [comment]
WOOOO!!!!!!!!!!!

**Comment by joost** [comment]
:+1:

**Comment by aanand** [comment]
LGTM

**Comment by dnephin** [comment]
LGTM!

**Comment by aanand** [comment]
giphy

**Comment by jonbrouse** [comment]
Thank you! 
![Celebrate]
(http://reactiongif.org/wp-content/uploads/GIF/2014/08/GIF-Barney-Stinson-celebrate-Champagne-excited-How-I-Met-Your-Mother-joy-jump-jump-for-joy-jumping-party-hard-pop-spray-woo-hoo-GIF.gif)

![OMG THANKYOU]
(http://i.minus.com/iknnd1vL0Ecww.gif)

**Comment by wokim** [comment]
:+1:

**Comment by stevenjack** [comment]
Great to see this finally in!

party

**Comment by zkilgore** [comment]
:+1::+1::+1::+1::+1::+1::+1::+1::+1::+1::+1:

**Comment by jasonferrier** [comment]
How about adding it to the [docker-compose.yml documentation](https://docs.docker.com/compose/yml/)? It took me a while to think of searching the Issues and PR's on here to figure out that it is possible...

**Comment by KyleJamesWalker** [comment]
Oh you beat me to it, can't wait for it in the main release! :P

**Comment by joost** [comment]
Docker Compose v1.3.0rc2 Homebrew Formula: https://gist.github.com/joost/20934650f68dd561fc35

**Comment by aanand** [comment]
Please note that Homebrew isn't an official installation method! Head here to install from binaries or pip: https://github.com/docker/compose/releases/tag/1.3.0rc2

**Pull request merged**